alarms/consumer/consumer.go

```diff
@@ -0,0 +1,255 @@
+			}
+		}
+		if msg.Name == "assignee_id" {
+			if msg.StringValue != nil {
+				assigneeID = *msg.StringValue
+			}
+		}
+		if msg.Name == "metadata" {
+			if msg.DataValue != nil {
+				data, err := json.Marshal(msg.DataValue)
+				if err == nil {
+					var payload map[string]interface{}
+					if err := json.Unmarshal(data, &payload); err == nil {
+						metadata = payload
+					}
+				}
+			}
+		}
+	}
+
+	a := alarms.Alarm{
+		RuleID:      ruleID,
+		DomainID:    domainID,
+		ChannelID:   msgs[0].Channel,
+		ThingID:     msgs[0].Publisher,
+		Subtopic:    msgs[0].Subtopic,
+		Status:      alarms.ActiveStatus,
+		Measurement: measurement,
+		Value:       value,
+		Unit:        unit,
+		Threshold:   threshold,
+		Cause:       cause,
+		Severity:    severity,
+		AssigneeID:  assigneeID,
+		CreatedAt:   time.Now(),
+		Metadata:    metadata,
+	}
+
+	if err := a.Validate(); err != nil {
+		return err
+	}
+
+	return c.svc.CreateAlarm(ctx, a)
+}
+
+func (c consumer) saveJSON(ctx context.Context, msgs smqjson.Messages) error {
+	var (
+		ruleID, domainID, measurement, value, unit, threshold, cause, assigneeID string
+		severity                                                                 uint8
+		metadata                                                                 map[string]interface{}
+	)
+
+	for _, msg := range msgs.Data {
+		if getString(msg.Payload, "rule_id") != "" {
+			ruleID = getString(msg.Payload, "rule_id")
+		}
+		if getString(msg.Payload, "domain_id") != "" {
+			domainID = getString(msg.Payload, "domain_id")
+		}
+		if getString(msg.Payload, "measurement") != "" {
+			measurement = getString(msg.Payload, "measurement")
+		}
+		if getString(msg.Payload, "value") != "" {
+			value = getString(msg.Payload, "value")
+		}
+		if getString(msg.Payload, "unit") != "" {
+			unit = getString(msg.Payload, "unit")
+		}
+		if getString(msg.Payload, "threshold") != "" {
+			threshold = getString(msg.Payload, "threshold")
+		}
+		if getString(msg.Payload, "cause") != "" {
+			cause = getString(msg.Payload, "cause")
+		}
+		if getUint8(msg.Payload, "severity") != nil {
+			s := getUint8(msg.Payload, "severity")
+			severity = *s
+		}
+		if getString(msg.Payload, "assignee_id") != "" {
+			assigneeID = getString(msg.Payload, "assignee_id")
+		}
+		if getString(msg.Payload, "metadata") != "" {
+			data, err := json.Marshal(getString(msg.Payload, "metadata"))
+			if err == nil {
+				var payload map[string]interface{}
+				if err := json.Unmarshal(data, &payload); err == nil {
+					metadata = payload
+				}
+			}
+		}
+	}
+
+	a := alarms.Alarm{
+		RuleID:      ruleID,
+		DomainID:    domainID,
+		ChannelID:   msgs.Data[0].Channel,
+		ThingID:     msgs.Data[0].Publisher,
+		Subtopic:    msgs.Data[0].Subtopic,
+		Status:      alarms.ActiveStatus,
+		Measurement: measurement,
+		Value:       value,
+		Unit:        unit,
+		Threshold:   threshold,
+		Cause:       cause,
+		Severity:    severity,
+		AssigneeID:  assigneeID,
+		CreatedAt:   time.Now(),
```

Comment:
I think this timestamp should ideally be provided by the rule engine. We could also consider adding another column to store the time when the entry was inserted into the table. What do you think?

Reply:
I agree. For the time being, let's rely on RE-sent timestamp.